### PR TITLE
Stop including flight block source in json

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The idea here is to remove the `method-source` gem that we aren't using anymore (the idea was originally to use it to stash away `*flight` blocks so that they could be loaded from the API be we decided to download the cask source instead) and that was actually causing some problems recently in https://github.com/Homebrew/brew/issues/14541. ~With that I think we should be able to remove the load from API code from the `#to_h*` methods which should help simplify the code a bit and maybe even allow us to avoid regression like https://github.com/Homebrew/brew/pull/14609 (still testing this part).~

~I also added a simple test showing that a subset of the Cask stanzas should be loadable from the source and the API and produce the same `#to_h` output.~

The idea is now for `*flight` blocks to just exist as an artifact name to null hash in the JSON. They are just there to indicate there presence and nothing more.

Update: The JSON generation using casks loaded from the API doesn't return the desired results but the removal of the flight block source in JSON is still ready to go.

### CODE REVIEW

**Comment by Bo98** [comment]
I agree embedding source in JSON should go, though backwards compatibility is starting to become a concern.

Though at least it seems like the API loader handles it gracefully in this case.

File: Library/Homebrew/cask/cask.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:300]: Could do similarly for `languages`

**Review by MikeMcQuaid — APPROVED** [review]
@apainintheneck This looks good to me. If you could fix it up ASAPish that'd be great as I think it makes sense to block 4.0.0 until it's done and I want to land 4.0.0 as 🔜 as possible.

**Comment by MikeMcQuaid** [comment]
> I'm still working on benchmarks

FYI: I think you can omit this for now. We can always make this faster afterwards.

**Comment by Bo98** [comment]
:+1: if this gets in for 4.0.0.

I suppose if we want to be careful, ideally the code that reads the data is removed now (before 4.0.0) and the code that generates the API data is removed after the default auto-update interval of 24 hours (this part doesn't need a tag).

File: Library/Homebrew/cask/cask_loader.rb
Comment (Rylan12) [inline Library/Homebrew/cask/cask_loader.rb:304]:
```suggestion
            send(key, *artifact[key]) if artifact[key].present?
```

Is this needed?

Comment (apainintheneck) [reply Library/Homebrew/cask/cask_loader.rb:304]: Why would that be needed? You can't reach this code if you use a flight stanza.

File: Library/Homebrew/cask/cask.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:300]: I think that including the languages in the JSON still is useful for users even if we don't use it.

**Review by apainintheneck — COMMENTED** [review]
If we're really in a hurry to get to `4.0.0` (why?) then I think it makes sense to split this up into two PRs. I'm confident in the removal of the `method_source` gem and surrounding code and even though that is technically a breaking change I doubt anyone used it in the first place.

Me playing around with removing the API code from the `#to_h*` methods is something I'm much less confident in. If I don't feel comfortable with that tonight, I'll just break that into another PR and work on it at my own pace. The goal with that is to see if they can be removed without breaking/changing the normal behavior so that can land whenever. It shouldn't need to wait for a major/minor release.

File: Library/Homebrew/cask/cask_loader.rb
Comment (apainintheneck) [reply Library/Homebrew/cask/cask_loader.rb:304]: You shouldn't make it past the following code if you have a flight artifact.

https://github.com/Homebrew/brew/blob/2a417f14163aacfb4ffcb754fe9dea71c46f6ad0/Library/Homebrew/cask/cask_loader.rb#L233-L238

**Comment by apainintheneck** [comment]
> 👍 if this gets in for 4.0.0.
> 
> I suppose if we want to be careful, ideally the code that reads the data is removed now (before 4.0.0) and the code that generates the API data is removed after the default auto-update interval of 24 hours (this part doesn't need a tag).

There is no code that is currently reading this data right now.

**Comment by Bo98** [comment]
That makes sense, thanks. I was mainly concerned about `instance_eval` but I see now that it is dead code.

In that case, it seems like everything here is on the API generation side.

**Comment by apainintheneck** [comment]
Looks like the `#to_h*` methods don't work for all casks in `homebrew/cask`. I'll revert the last two commits and just go with the remove the flight blocks code.

<details>
<summary>Script</summary>

```rb
# For testing how cask json generation works after
# a Cask is loaded from the API.

require "cask"

CASK_TOKENS = Tap.new("homebrew", "cask").cask_tokens.freeze

class Stats
  attr_accessor :matches, :mismatched
  attr_reader :method

  def initialize(method)
    @matches = 0
    @mismatched = {}
    @method = method
  end
end

def to_json(hash)
  JSON.pretty_generate(hash)
end

def hash_diff(lhs, rhs)
  (lhs.keys | rhs.keys).reject do |key|
    lhs[key].to_s == rhs[key].to_s
  end
end

to_h_stats = Stats.new :to_h
to_h_var_stats = Stats.new :to_hash_with_variations
total = 0

CASK_TOKENS.each do |token|
  total += 1
  print "." if total % 30 == 0

  local = Cask::CaskLoader.load(token)

  local_hash_var = local.to_hash_with_variations

  api = Cask::CaskLoader::FromAPILoader
    .new(token, from_json: JSON.parse(to_json(local_hash_var)))
    .load(config: nil)

  api_hash_var = api.to_hash_with_variations

  if to_json(local_hash_var) == to_json(api_hash_var)
    to_h_var_stats.matches += 1
  else
    to_h_var_stats.mismatched[token] = hash_diff(local_hash_var, api_hash_var)
  end

  local_hash = local.to_h
  api_hash = api.to_h
  if to_json(local_hash) == to_json(api_hash)
    to_h_stats.matches += 1
  else
    to_h_stats.mismatched[token] = hash_diff(local_hash, api_hash)
  end
end

puts

[
  to_h_stats,
  to_h_var_stats,
].each do |stats|
  puts "-----"
  puts "Cask##{stats.method} : #{stats.matches}/#{total}"

  counter = stats
    .mismatched
    .values
    .flatten
    .each_with_object(Hash.new(0)) do |key, obj|
      obj[key] += 1
    end

  puts
  puts "Mismatched keys:"
  counter.sort_by { |x| x[1] }.each { |x| p x }
end
puts "-----"
```

</details>

```
-----
Cask#to_h : 3965/4154

Mismatched keys:
["auto_updates", 1]
["depends_on", 1]
["caveats", 17]
["artifacts", 20]
["tap", 154]
["tap_git_head", 154]
-----
Cask#to_hash_with_variations : 3490/4154

Mismatched keys:
["auto_updates", 1]
["depends_on", 1]
["caveats", 17]
["artifacts", 27]
["tap", 154]
["tap_git_head", 154]
["variations", 533]
-----
```

File: Library/Homebrew/cask/cask.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/cask/cask.rb:316]: @apainintheneck Given no users ever requested it and it increases the size of the JSON: I disagree.

**Comment by MikeMcQuaid** [comment]
Thanks @apainintheneck!

**Pull request merged**
